designs/0026-stan-regression-syntax.md

```diff
@@ -0,0 +1,256 @@
+
+There's really nothing to do more with the population intercept and slope. For everything else, a basic syntax that would be suitable for Stan would be:
+
+```
+covariate .* (parameters | group_idx)
+```
+
+where:
+
+1. `group_idx` is the group index of the responses. It would be an integer array of length `N` in the range `[1, G]` assuming `G` groups.
+2. `parameters` are the group level coefficients. It would be a length `G` vector-like variable.
+3. `covariate` is an optional, length `N` vector-like set of covariates.
+
+In the simplest case where `group_idx` is a 1d array of integers of length N, `parameters` is a vector of length `G`, and `covariate` is a length `N` vector of coefficients this would be equivalent to the Stan code here:
+
+```
+covariate .* parameters[group_idx]
+```
+
+This is not anything beyond what is already possible, but with a few variations on the syntax it can be.
+
+First, this syntax can support multiple-group indexing. Instead of writing loops, or doing an inefficient `covariate .* diagonal(parameters[group1, group2])`, or refactoring the two groups into one, the multiple group membership could be handled by:
+
+```
+covariate .* (parameters | group1, group2)
+```
+
+where `parameters` is now a 2d variable. To get the design matrices, the `parameters` variable could be removed.
+
+So if the following returns a length `N` vector:
+```
+covariate .* (parameters | group)
+```
+
+Then this would return an `N` by `G` design matrix:
+```
+design_matrix dmat = covariate .* (1 | group)
+```
+
+And then this design matrix could be cast to a `sparse_matrix` (once these are in Stan):
+```
+sparse_matrix smat = mat;
+```
+
+With the sparse matrix, the matrix vector product could be evaluated directly to get the equivalent of `covariate .* (parameters | group)`:
+
+```
+smat * parameters
+```
+
+But the design matrix could also get a `operator()` so that this was equivalent to:
+
+```
+dmat(parameters)
+```
+
+It might seem awkward to do this with call syntax, but this might be clearer for multiple group design matrices. For instance:
+
+```
+design_matrix dmat2 = (1 | group1, group2);
+```
+
+In this case `coefficients_matrix` would need compressed to a vector to make the sparse matrix vector product make sense, whereas `dmat2(coefficients_matrix)` can mean the effective matrix vector product without worrying about flattening `coefficients_matrix`. An extra function `flatten` could be provided to do this reordering if necessary.
+
+With this syntax, the regression from above could be rewritten:
+```
+vector[N] mu = intercept + slope * x +
+    (group1_intercepts | group1_idx) + x .* (group1_slopes | group1_idx) +
+    (group12_intercepts | group1_idx, group2_idx) +
+    x .* (group12_slopes | group1_idx, group2_idx);
```

Comment:
Looking at this I'm not really sure exactly how we'd optimize this on the backend. I kind of like something similar to your original proposal for a `hiearchy()` function for something like 

```
vector[N] mu = intercept + slope * x + hierarchy(f, group1_idx, group2_idx, group12_intercepts, x, group12_slopes)
```

Where `f` is a function with a signature like `vector f(real intercept, matrix x, vector slope)` that would be applied to each group (in this case the group is the combo of groups 1 and 2).

I wonder if we could do anything useful if we have a `group_index` type in the stan language? Or could we give matrices an index? Maybe something like 

```
group_index[N] group_idx1;
group_index[N] group_idx2;
// Anything after first two dims for a matrix are indices?
matrix[N, M, group_idx1, group_idx2] X;
// Or maybe a tag?
matrix[N, M] X index(group_idx1, group_idx2);
```
Though I think the issue with what I have here is that we would have to pay for a sort / unsort if the data was not presorted so we could iterate over the groups nicely

Reply:
> I'm not really sure exactly how we'd optimize this on the backend

I imagined this would be something recognizable by the compiler. Like, "oh, I have a bunch of design matrices, better lump them together and turn them into a big `hierarchy` statement".

But also I think @t4c1 likes expressions and I could see that working too, so it kinda all comes together at runtime. As long as it gets into a big `hierarchy`-like statement it's good.

> group_index type in the stan language
> we would have to pay for a sort / unsort if the data

Is the goal here to express a sparseness or to express the ordering of the matrix to make accesses efficient in some way?